Breeding-Insight/bi-api
# BI-1881 - Store Parent UUIDs in Germplasm AdditionalInfo

# Description
**Story:** https://breedinginsight.atlassian.net/browse/BI-1881

Previously, `BrAPIGermplasmDAO::processGermplasmForDisplay` required parent `BrAPIGermplasm` objects to be available in order to add the `pedigreeByUUID` and `pedigreeByName` keys to `additionalInfo` on each `BrAPIGermplasm`. This required some hacks, [passing all germplasm in the program for example](https://github.com/Breeding-Insight/bi-api/blob/09ccfd5fa20dd199d7b00cd8f45516e3d2fa0b14/src/main/java/org/breedinginsight/brapi/v2/dao/BrAPIGermplasmDAO.java#L295).

To make the code more efficient and avoid having to pass all program germplasm into `processGermplasmForDisplay`, I made changes to
- store male/female parent UUIDs on each `BrAPIGermplasm` in `additionalInfo` on create,
- read the parent UUIDs in `processGermplasmForDisplay` to build the required pedigree string ("<femaleParentUUID>/<maleParentUUID>"),
- wrote a SQL script to update Germplasm records in BreedBase to store the parental UUIDs in `additionalInfo`, that's in [this PR on breedbase_configs](https://github.com/Breeding-Insight/breedbase_configs/pull/32).

>Note: because the BrAPI Java TestServer (BJTS) stores additional info as byte arrays of serialized Java objects, it was deemed impractical to make the change in SQL. For the 1 program using the BJTS in production, there are no germplasm records so there is no data to be upgraded. The non-production BJTS instances will be wiped out, it's on [the v0.8.1 deploy checklist](https://breedinginsight.atlassian.net/l/cp/9KQLE3Np).


# Dependencies
If you checkout this branch and run it against BJTS or BreedBase with existing germplasm records, it will break. In the case of BJTS, the only option currently is to wipe out the database. For BreedBase, germplasm records can be upgraded in place by running `breedbase-deploy.sh` from the `bug/BI-1881-mlm483` branch. [See this PR for details](https://github.com/Breeding-Insight/breedbase_configs/pull/32).

# Testing

Steps 1-10 focus on BreedBase, step 11 is for BJTS.

<details>
<summary><strong>germplasm_with_additional_info</strong> (click to expand)</summary>

```sql
-- Germplasm with additional info.
SELECT 
    germ.stock_id, germ.uniquename AS "stock.uniquename", cvterm.name AS "cvterm.name", stockprop.stockprop_id, stockprop.value, stockprop.rank
FROM 
    public.stock germ
    LEFT JOIN public.stockprop ON stockprop.stock_id = germ.stock_id
    LEFT JOIN cvterm ON cvterm.cvterm_id = stockprop.type_id
WHERE 
    germ.stock_id IN
    (
        SELECT
            stock.stock_id
        FROM 
            stock
            JOIN
            cvterm cvt ON cvt.cvterm_id = stock.type_id AND cvt.name::text = 'accession'::text
    )
    AND cvterm.name = 'stock_additional_info'
    -- Uncomment if you only want to see germplasm from the attached file.
    -- AND germ.uniquename LIKE '1881-%'
ORDER BY germ.uniquename, cvterm.name, stockprop.rank
```

</details>

1. Checkout the `develop` branch for both `bi-api` and `breedbase_configs`.
2. Wipe out BJTS and BreedBase (optional, but makes testing cleaner). I like to delete all local containers and volumes.
3. Start up BreedBase (with `breedbase_deploy.sh`) and start up all the services needed for DeltaBreed, then run `bi-api` and `bi-web`.
4. Create a new program stored in BreedBase and upload germplasm with various pedigree permutations (example file attached).
5. Check the additionalInfo doesn't contain "maleParentUUID" or "femaleParentUUID" fields with the germplasm_with_additional_info SQL query above. If those keys are present, you are probably running the `bug/BI-1881-mlm483` branch of `bi-api`. Oops, start over!
6. Stop `bi-api` and checkout `bug/BI-1881-mlm483`. If you restart before performing the next step, you'll see some exceptions saying "Germplasm has a female parent but femaleParentUUID is missing.".
7. Checkout the `bug/BI-1881-mlm483` branch of `breedbase_configs` and run `breedbase_deploy.sh` (without the `--reinstall` flag). The flyway docker container will be pulled, start up, wait for the breedbase_db container to be healthy and then run the migrations in the ./flyway/sql directory.
8. You can confirm that the data has been updated with the germplasm_with_additional_info SQL above.
9. Start `bi-api` and ensure it runs without the exception and germplasm list, detail and pedigree views work as expected.
10. Upload more germplasm, and use steps 8. and 9. to ensure the running code stores new germplasm with the parental UUIDs correctly.
11. Create a program in the BJTS and upload germplasm and ensure the upload and list, detail and pedigree views work as expected. You can use the `brapi/v2/germplasm` endpoint as well to see that the "femaleParentUUID" and/or "maleParentUUID" fields are present in `additionalInfo`.

# Checklist:

- [x] I have run TAF: https://github.com/Breeding-Insight/taf/actions/runs/6253277974
